sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -171,6 +172,10 @@ object ColumnPruning extends Rule[LogicalPlan] {
 
       Project(substitutedProjection, child)
 
+    // add a project which is blocked by Generate
+    case p @ pushBelowGenerate(newChild) =>
```

Comment:
Just look for Project on top of generate instead of attempting to walk the query tree.  You can assume other rules will push projects down to you.